lib/single_sign_on.rb

```diff
@@ -83,6 +88,18 @@ def self.parse(payload, sso_secret = nil)
     sso
   end
 
+  def self.provider_secret(return_sso_url)
+    provider_secrets = SiteSetting.sso_provider_secrets.split(/[\|,\n]/)
+    provider_secrets_hash = Hash[*provider_secrets]
+    return_url = URI.parse(return_sso_url).host
+
+    secret = provider_secrets_hash.select do |domain, _|
+      domain_regex = Regexp.escape(domain).gsub("\\*", '\S*')
```

Comment:
Is there a bug here ... Say we allow. `*.discourse.org` will `bad-domain.discourse.org.evil.com` slip through? maybe extract some of this logic into a service object of sorts and add some tests for the security aspects?